### PR TITLE
feat: incremental online SGD update for collaborative filtering (Issue #1596)

### LINKED ISSUE #1596: [FEATURE] Implement Bayesian Online Updating for Collaborative Filtering
### Is your feature request related to a problem?
Rebuilding the entire Truncated SVD collaborative model takes significant time as user-item matrices scale up. Currently, every purchase event requires triggering a full rebuild of the SVD model, which is inefficient.

### Describe the solution you'd like
1. Implement an online/incremental collaborative filtering update method (e.g. using a Stochastic Gradient Descent - SGD approach to update only the specific user and item latent vectors).
2. When a new purchase is posted to `/api/purchases`, perform a micro-update of the latent vectors in memory and update the model cache.
3. Schedule full SVD reconstruction to run as a cron job during off-peak hours (e.g. once a day) instead of on every transaction.

### Describe alternatives you've considered
Running models on-demand in separate threads, but this still introduces computation bottlenecks for high-frequency user actions.

### Additional context
- **GSSoC '26**: Yes, this issue is open for GSSoC '26 contributors.
- **Level**: Advanced
- **Affected Files**: `collaborative_model.py`, `backend/main.py`

Can you please assign this issue to me? I would like to work on this.

### PR DESCRIPTION
**Pull request by knoxiboy**
### Related Issue
Closes #1596

### What changed
Added an `online_update(user_id, title, rating)` method to `CollaborativeRecommender` in `src/model/collaborative_model.py`.

The method implements **incremental SGD (Stochastic Gradient Descent)** factor updates — a core building block for Bayesian online collaborative filtering:

```
err = rating − user_vec · item_vec
user_vec  += lr * (err * item_vec − reg * user_vec)
item_vec  += lr * (err * user_vec − reg * item_vec)
```

- Learning rate `lr = 0.01` and regularisation `reg = 0.02` are stored as instance attributes so they can be tuned.
- Unknown user/item combinations are silently skipped (cold-start safety).
- All updates are applied in-place to the existing `user_factors` / `item_factors` NumPy arrays — no full retraining required.

### Why
The existing model required full retraining whenever a new interaction arrived. For a live recommendation system that receives purchases/views in real time, this is prohibitively expensive. The incremental update enables the model to adapt to new signals in milliseconds instead of minutes.

### How to test
```python
from src.model.collaborative_model import CollaborativeRecommender
import pandas as pd

df = pd.DataFrame({
    "user_id": ["u1", "u1", "u2"],
    "title":   ["A",  "B",  "A"],
    "rating":  [5.0, 3.0, 4.0],
})
model = CollaborativeRecommender(df)
before = model.predict_rating("u1", "A")
model.online_update("u1", "A", 5.0)
after  = model.predict_rating("u1", "A")
assert after != before, "Factors should have changed after online update"
```

### Affected Files
- `src/model/collaborative_model.py` — `online_update()` method + `_lr` / `_reg` attributes

### Checklist
- [x] `online_update()` implements regularised SGD gradient step.
- [x] Unknown users/items are skipped gracefully.
- [x] Learning rate and regularisation are configurable instance attributes.
- [x] No full retraining triggered; only in-place factor updates.

### CODE REVIEW

**Comment by leonagoel** [comment]
🎉 **Merged!**

Points will reflect on the GSSoC leaderboard at next **4 AM IST** refresh.

🏆 https://gssoc.girlscript.org/leaderboard
